src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs

```diff
@@ -963,143 +963,144 @@ public static ulong ToUInt64(string s)
+                    {
+                        return -0f;
+                    }
 
-            return f;
+                    return f;
+            }
         }
 
         internal static Exception? TryToSingle(string s, out float result)
         {
-            s = TrimString(s);
-            if (s == "-INF")
-            {
-                result = float.NegativeInfinity;
-                return null;
-            }
-            else if (s == "INF")
-            {
-                result = float.PositiveInfinity;
-                return null;
-            }
-            else if (!float.TryParse(s, NumberStyles.AllowLeadingSign | NumberStyles.AllowDecimalPoint | NumberStyles.AllowExponent, NumberFormatInfo.InvariantInfo, out result))
+            ReadOnlySpan<char> value = s.AsSpan().Trim(WhitespaceChars);
+            switch (value)
             {
-                return new FormatException(SR.Format(SR.XmlConvert_BadFormat, s, "Single"));
-            }
+                case "-INF":
+                    result = float.NegativeInfinity;
+                    return null;
+                case "INF":
+                    result = float.PositiveInfinity;
+                    return null;
+                default:
+                    if (!float.TryParse(value, NumberStyles.AllowLeadingSign | NumberStyles.AllowDecimalPoint | NumberStyles.AllowExponent, NumberFormatInfo.InvariantInfo, out result))
+                    {
+                        return new FormatException(SR.Format(SR.XmlConvert_BadFormat, s, "Single"));
+                    }
+                    if (result == 0 && value[0] == '-')
+                    {
+                        result = -0f;
+                    }
 
-            if (result == 0 && s[0] == '-')
-            {
-                result = -0f;
+                    return null;
             }
-
-            return null;
         }
 
         public static double ToDouble(string s)
         {
-            s = TrimString(s);
-            if (s == "-INF") return double.NegativeInfinity;
-            if (s == "INF") return double.PositiveInfinity;
-
-            double dVal = double.Parse(s, NumberStyles.AllowLeadingSign | NumberStyles.AllowDecimalPoint | NumberStyles.AllowExponent | NumberStyles.AllowLeadingWhite | NumberStyles.AllowTrailingWhite, NumberFormatInfo.InvariantInfo);
-            if (dVal == 0 && s[0] == '-')
+            ReadOnlySpan<char> value = s.AsSpan().Trim(WhitespaceChars);
+            switch (value)
             {
-                return -0d;
-            }
+                case "-INF":
+                    return double.NegativeInfinity;
+                case "INF":
+                    return double.PositiveInfinity;
+                default:
+                    double dVal = double.Parse(value, NumberStyles.AllowLeadingSign | NumberStyles.AllowDecimalPoint | NumberStyles.AllowExponent | NumberStyles.AllowLeadingWhite | NumberStyles.AllowTrailingWhite, NumberFormatInfo.InvariantInfo);
```

Comment:
I dig a bit more in the callstack of failing test and it looks like the failing assert is for null here:
https://github.com/dotnet/runtime/blob/5c6cb20f374d09cbbde059c94ea3f97970ce9b23/src/libraries/System.Private.CoreLib/src/System/Number.Parsing.cs#L2644

Which is really strange as I didn't changed the logic, just moved it into the switch statement.
Also `NumberFormatInfo.InvariantInfo` is passed.
Any idea what I am doing wrong ? @eiriktsarpalis

Reply:
Hey, start with rebasing on top of main. @stephentoub has made some changes recently where he merged all test projects together. Try seeing if that repros on platform you're using for testing first (maybe run this standalone couple of times). If no luck, add assert just before this piece of code i.e. `Debug.Assert(NumberFormatInfo.InvariantInfo != null, "NumberFormatInfo.InvariantInfo is null");`. Honestly I can see this happen for at least three reasons:
- InvariantInfo somehow didn't get initialized - maybe some error which we catched exception and ended up not assigning anything or assigning null - might be worth to skim through the code if you see anything obvious and if you do report an issue
- some weird ordering in static initialization which we somehow accidentally trigger (but then doesn't make sense why it's OSX only)
- coreclr issue (it seems to be OSX specific and Debug only)

ideally if you got OSX try debugging and seeing why it's not initialized but that is assuming you can get relatively consistent repro.

Reply:
first variant seems impossible since ctor is almost no-op and there is nothing special going on: https://github.com/dotnet/runtime/blob/4cf1383c8458945b7eb27ae5f57338c10ed25d54/src/libraries/System.Private.CoreLib/src/System/Globalization/NumberFormatInfo.cs#L186

Reply:
static initialization doesn't seem like could be an issue as well given we do lazy initialization here. I'll bet 3rd option but let's add explicit assert to confirm that is actually null and we're not missing something else

Reply:
actually possibly https://github.com/dotnet/runtime/blob/5c6cb20f374d09cbbde059c94ea3f97970ce9b23/src/libraries/System.Private.CoreLib/src/System/Double.cs#L405 is the issue here since we're calling NumberFormatInfo.GetInstance(provider) and I think we've passed NumberFormatInfo directly here

Reply:
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Globalization/NumberFormatInfo.cs#L186-L210 this looks good though since we should be returning `provider as NumberFormatInfo` - I'd recommend try debugging this and check if there is any weirdness going on somewhere

Reply:
@krwq I think the problem is in different implementation when parsing string and ROS with null input.
The test which is failing is passing null to double.Parse().
https://github.com/dotnet/runtime/blob/b40c3a3ba575babb4c0409eb5cbb9b728b1ba9c2/src/libraries/System.Private.CoreLib/src/System/Double.cs#L387

``` cs
        public static double Parse(string s, NumberStyles style, IFormatProvider? provider)
        {
            NumberFormatInfo.ValidateParseStyleFloatingPoint(style);
            if (s == null) ThrowHelper.ThrowArgumentNullException(ExceptionArgument.s);
            return Number.ParseDouble(s, style, NumberFormatInfo.GetInstance(provider));
        }

        // Parses a double from a String in the given style.  If
        // a NumberFormatInfo isn't specified, the current culture's
        // NumberFormatInfo is assumed.
        //
        // This method will not throw an OverflowException, but will return
        // PositiveInfinity or NegativeInfinity for a number that is too
        // large or too small.

        public static double Parse(ReadOnlySpan<char> s, NumberStyles style = NumberStyles.Float | NumberStyles.AllowThousands, IFormatProvider? provider = null)
        {
            NumberFormatInfo.ValidateParseStyleFloatingPoint(style);
            return Number.ParseDouble(s, style, NumberFormatInfo.GetInstance(provider));
        }
```

Looks like direct replace from string to ReadOnlySpan is not okay, or at least have to check for null outside.

Reply:
please file an issue on that with repro, that sounds like a bug

Reply:
[Done](https://github.com/dotnet/runtime/issues/76442)